### PR TITLE
Feat course category

### PR DESCRIPTION
**Pull request by nskalu**
added the endpoint for add, update, delete and retrieve of course categories.

### CODE REVIEW

**Comment by tayormi** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 16
           
```


See the complete overview on [Codacy](https://app.codacy.com/gh/codeclannigeria/codeclannigeria-backend/pullRequest?prid=5693791&bid=18633753)

File: src/categories/models/category.entity.ts [hunk 1]
```diff
@@ -0,0 +1,16 @@
+import { columnSize } from '@shared/constants';
+import { BaseEntity } from '@shared/models/base.entity';
+import { index, prop } from '@typegoose/typegoose';
```
Comment (tayormi) [inline src/categories/models/category.entity.ts:3]: Codacy Issue found: ['index' is defined but never used.](https://app.codacy.com/gh/codeclannigeria/codeclannigeria-backend/pullRequest?prid=5693791)

File: README.md [hunk 2]
```diff
@@ -1,3 +1,34 @@
...
+
+## How You Come In
+
+To get a complete documentation of the source code and the architectural structure of the application, **[click here](https://codeclannigeria.github.io/codeclannigeria-backend/docs)**.
+
+Feel free to put your skills to the test. Here is a suggested workflow for you to follow:
+**Note: This is applicable to codeclannigeria community members only.**
+
+1. Fork this repository into your local computer
+2. Check the issues pane for an issue that needs to be solved
```
Comment (tayormi) [inline README.md:25]: Codacy Issue found: [[list-item-indent] Incorrect list-item indent: add 1 space](https://app.codacy.com/gh/codeclannigeria/codeclannigeria-backend/pullRequest?prid=5693791)

File: README.md [hunk 3]
```diff
@@ -1,3 +1,34 @@
...
+3. Find an issue that matches your current skill level and solve it
+4. Make your pull request to the `develop` branch
```
Comment (tayormi) [inline README.md:27]: Codacy Issue found: [[list-item-indent] Incorrect list-item indent: add 1 space](https://app.codacy.com/gh/codeclannigeria/codeclannigeria-backend/pullRequest?prid=5693791)

File: src/categories/categories.controller.ts [hunk 4]
```diff
@@ -0,0 +1,119 @@
+import {
+    Body,
+    ConflictException,
+    Controller,
+    HttpCode,
+    HttpStatus,
+    Post,
+    UseGuards,
```
Comment (tayormi) [inline src/categories/categories.controller.ts:8]: Codacy Issue found: ['UseGuards' is defined but never used.](https://app.codacy.com/gh/codeclannigeria/codeclannigeria-backend/pullRequest?prid=5693791)

File: README.md [hunk 5]
```diff
@@ -1,3 +1,34 @@
...
+5. Wait for some feedback.
```
Comment (tayormi) [inline README.md:28]: Codacy Issue found: [[list-item-indent] Incorrect list-item indent: add 1 space](https://app.codacy.com/gh/codeclannigeria/codeclannigeria-backend/pullRequest?prid=5693791)

File: README.md [hunk 6]
```diff
@@ -1,3 +1,34 @@
...
+[![Codacy Badge](https://api.codacy.com/project/badge/Grade/34b7b5288f2f451eb7224cdac4197041)](https://app.codacy.com/gh/codeclannigeria/codeclannigeria-backend?utm_source=github.com&utm_medium=referral&utm_content=codeclannigeria/codeclannigeria-backend&utm_campaign=Badge_Grade_Dashboard) [![build status](https://github.com/coryrylan/angular-github-actions/workflows/Build/badge.svg)](https://github.com/codeclannigeria/codeclannigeria-backend/actions)
+
 The official website for code clan nigeria community
+We are a community comprised of future innovators.
+Front-end developers, DevOps engineers, full stack developers, technical
+writers, software engineers.
+This community was created out of a desire to provide ongoing support
+to you wherever stage you are in your coding career.
+
+# This is codeclannigeria.dev
```
Comment (tayormi) [inline README.md:12]: Codacy Issue found: [[no-multiple-toplevel-headings] Don’t use multiple top level headings (1:1)](https://app.codacy.com/gh/codeclannigeria/codeclannigeria-backend/pullRequest?prid=5693791)

File: src/categories/categories.service.ts [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+    super(categoryEntity);
+  }
+  async InsertCategory(categoryname: string){
+    const newCat = new this.categoryEntity({category:categoryname});
+   var result = await newCat.save();
+   return result.id;
+  }
+
+  async GetAllCategories(){
+   var categories = await this.categoryEntity.find({isDeleted:false}).exec();
```
Comment (tayormi) [inline src/categories/categories.service.ts:24]: Codacy Issue found: [Unexpected var, use let or const instead.](https://app.codacy.com/gh/codeclannigeria/codeclannigeria-backend/pullRequest?prid=5693791)

File: README.md [hunk 8]
```diff
@@ -1,3 +1,34 @@
...
+## How You Come In
+
+To get a complete documentation of the source code and the architectural structure of the application, **[click here](https://codeclannigeria.github.io/codeclannigeria-backend/docs)**.
+
+Feel free to put your skills to the test. Here is a suggested workflow for you to follow:
+**Note: This is applicable to codeclannigeria community members only.**
+
+1. Fork this repository into your local computer
+2. Check the issues pane for an issue that needs to be solved
+3. Find an issue that matches your current skill level and solve it
```
Comment (tayormi) [inline README.md:26]: Codacy Issue found: [[list-item-indent] Incorrect list-item indent: add 1 space](https://app.codacy.com/gh/codeclannigeria/codeclannigeria-backend/pullRequest?prid=5693791)

File: src/categories/categories.service.ts [hunk 9]
```diff
@@ -0,0 +1,27 @@
+import { Injectable, Scope } from '@nestjs/common';
+import { InjectModel } from '@nestjs/mongoose';
+import { BaseService } from '@shared/services';
+import { ReturnModelType } from '@typegoose/typegoose';
+
+import { Category } from './models/category.entity';
+import { CategoryDto } from './models/dto/category.dto';
```
Comment (tayormi) [inline src/categories/categories.service.ts:7]: Codacy Issue found: ['CategoryDto' is defined but never used.](https://app.codacy.com/gh/codeclannigeria/codeclannigeria-backend/pullRequest?prid=5693791)

File: README.md [hunk 10]
```diff
@@ -1,3 +1,34 @@
...
+4. Make your pull request to the `develop` branch
+5. Wait for some feedback.
+6. Repeat the process
```
Comment (tayormi) [inline README.md:29]: Codacy Issue found: [[list-item-indent] Incorrect list-item indent: add 1 space](https://app.codacy.com/gh/codeclannigeria/codeclannigeria-backend/pullRequest?prid=5693791)

File: src/categories/categories.controller.ts [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+    ApiBadRequestResponse,
+    ApiNotFoundResponse
+  } from '@nestjs/swagger';
+  import { AbstractCrudController } from '@shared/base.controller';
+  import { plainToClass } from 'class-transformer';
+  
+  //import { JwtAuthGuard } from '../auth/guards/jwt-auth.guard';
+  import { CategoryDto } from './models/dto/category.dto';
+  //import { TrackDto } from './models/dto/tack.dto';
+  import { Category } from './models/category.entity';
```
Comment (tayormi) [inline src/categories/categories.controller.ts:30]: Codacy Issue found: ['Category' is defined but never used.](https://app.codacy.com/gh/codeclannigeria/codeclannigeria-backend/pullRequest?prid=5693791)

File: src/categories/categories.controller.ts [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+  import { CategoryService } from './categories.service';
+  import { PagedCategoryResDto } from './models/dto/paged-category.dto';
```
Comment (tayormi) [inline src/categories/categories.controller.ts:32]: Codacy Issue found: ['PagedCategoryResDto' is defined but never used.](https://app.codacy.com/gh/codeclannigeria/codeclannigeria-backend/pullRequest?prid=5693791)

File: src/categories/categories.controller.ts [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+    Logger
+  } from '@nestjs/common';
+  import {
+    ApiBearerAuth,
+    ApiTags,
+    ApiOkResponse,
+    ApiBadRequestResponse,
+    ApiNotFoundResponse
+  } from '@nestjs/swagger';
+  import { AbstractCrudController } from '@shared/base.controller';
```
Comment (tayormi) [inline src/categories/categories.controller.ts:24]: Codacy Issue found: ['AbstractCrudController' is defined but never used.](https://app.codacy.com/gh/codeclannigeria/codeclannigeria-backend/pullRequest?prid=5693791)

File: src/categories/categories.service.ts [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+export class CategoryService extends BaseService<Category> {
+  constructor(
+    @InjectModel(Category.modelName)
+    protected readonly categoryEntity: ReturnModelType<typeof Category>,
+  ) {
+    super(categoryEntity);
+  }
+  async InsertCategory(categoryname: string){
+    const newCat = new this.categoryEntity({category:categoryname});
+   var result = await newCat.save();
```
Comment (tayormi) [inline src/categories/categories.service.ts:19]: Codacy Issue found: [Unexpected var, use let or const instead.](https://app.codacy.com/gh/codeclannigeria/codeclannigeria-backend/pullRequest?prid=5693791)

File: README.md [hunk 15]
```diff
@@ -1,3 +1,34 @@
...
+stand for.
+
+## How You Come In
+
+To get a complete documentation of the source code and the architectural structure of the application, **[click here](https://codeclannigeria.github.io/codeclannigeria-backend/docs)**.
+
+Feel free to put your skills to the test. Here is a suggested workflow for you to follow:
+**Note: This is applicable to codeclannigeria community members only.**
+
+1. Fork this repository into your local computer
```
Comment (tayormi) [inline README.md:24]: Codacy Issue found: [[list-item-indent] Incorrect list-item indent: add 1 space](https://app.codacy.com/gh/codeclannigeria/codeclannigeria-backend/pullRequest?prid=5693791)

File: src/categories/categories.controller.ts [hunk 16]
```diff
@@ -0,0 +1,119 @@
+import {
+    Body,
+    ConflictException,
+    Controller,
+    HttpCode,
+    HttpStatus,
+    Post,
+    UseGuards,
+    Get,
+    Query,
```
Comment (tayormi) [inline src/categories/categories.controller.ts:10]: Codacy Issue found: ['Query' is defined but never used.](https://app.codacy.com/gh/codeclannigeria/codeclannigeria-backend/pullRequest?prid=5693791)

**Comment by ofuochi** [comment]
Hi @nskalu 
Check @tayormi reviews and fix accordingly

**Comment by tayormi** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 8
           
```


See the complete overview on [Codacy](https://app.codacy.com/gh/codeclannigeria/codeclannigeria-backend/pullRequest?prid=5693791&bid=18633753)

File: README.md (hunk 3 above)
Comment (tayormi) [inline README.md:27]: Codacy Issue found: [[list-item-indent] Incorrect list-item indent: add 1 space](https://app.codacy.com/gh/codeclannigeria/codeclannigeria-backend/pullRequest?prid=5693791)

File: README.md (hunk 2 above)
Comment (tayormi) [inline README.md:25]: Codacy Issue found: [[list-item-indent] Incorrect list-item indent: add 1 space](https://app.codacy.com/gh/codeclannigeria/codeclannigeria-backend/pullRequest?prid=5693791)

File: README.md (hunk 6 above)
Comment (tayormi) [inline README.md:12]: Codacy Issue found: [[no-multiple-toplevel-headings] Don’t use multiple top level headings (1:1)](https://app.codacy.com/gh/codeclannigeria/codeclannigeria-backend/pullRequest?prid=5693791)

File: README.md (hunk 5 above)
Comment (tayormi) [inline README.md:28]: Codacy Issue found: [[list-item-indent] Incorrect list-item indent: add 1 space](https://app.codacy.com/gh/codeclannigeria/codeclannigeria-backend/pullRequest?prid=5693791)

File: README.md (hunk 15 above)
Comment (tayormi) [inline README.md:24]: Codacy Issue found: [[list-item-indent] Incorrect list-item indent: add 1 space](https://app.codacy.com/gh/codeclannigeria/codeclannigeria-backend/pullRequest?prid=5693791)

File: README.md (hunk 10 above)
Comment (tayormi) [inline README.md:29]: Codacy Issue found: [[list-item-indent] Incorrect list-item indent: add 1 space](https://app.codacy.com/gh/codeclannigeria/codeclannigeria-backend/pullRequest?prid=5693791)

File: README.md (hunk 8 above)
Comment (tayormi) [inline README.md:26]: Codacy Issue found: [[list-item-indent] Incorrect list-item indent: add 1 space](https://app.codacy.com/gh/codeclannigeria/codeclannigeria-backend/pullRequest?prid=5693791)

File: src/categories/categories.service.ts [hunk 17]
```diff
@@ -0,0 +1,26 @@
...
```
Comment (tayormi) [inline src/categories/categories.service.ts:18]: Codacy Issue found: [Unexpected var, use let or const instead.](https://app.codacy.com/gh/codeclannigeria/codeclannigeria-backend/pullRequest?prid=5693791)

**Pull request merged**